### PR TITLE
Enable use of transmission techs in custom math

### PR DESCRIPTION
**Pull request by brynpickering**
Adds ability to reference transmission technologies by the tech name, which will grab the relevant `tech:node_to` concatenated techs. 

Summary of changes in this pull request:

* New expression helper function to get transmission techs, but also allow it to work in the `defined` where string helper function.
* New custom math to showcase it in action, creating a net import share constraint.
* Revoke option to use `-` in node names, replacing with `_` (will need baking into a YAML schema in due course). With `-` there was a parsing issue raised if directly referencing the tech name in the where/expression string (`-` is not part of the allowed string characters).

Reviewer checklist:

- [x] Test(s) added to cover contribution
- [ ] Documentation updated
- [ ] Changelog updated
- [ ] Coverage maintained or improved

### CODE REVIEW

File: doc/_static/custom_math/net_import_share.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    description: Limit upper bound on annual heat imports across two nodes in the model as a share of combined heat flows in both nodes.
+    equations:
+      - expression: net_import_share * sum(flow_out[techs=$transmission_techs, nodes=$node_group, carriers=$carrier], over=[nodes, techs, timesteps]) <= $total_energy_balance
+    sub_expressions:
+      total_energy_balance:
+        - expression: sum(flow_out[nodes=$node_group, carriers=$carrier], over=[nodes, techs, timesteps]) - sum(flow_in[nodes=$node_group, carriers=$carrier], over=[nodes, techs, timesteps])
+    slices:
+      transmission_techs:
+        - expression: get_transmission_techs(test_transmission_heat)
+      node_group:
```
Comment (sjpfenninger) [inline doc/_static/custom_math/net_import_share.yaml:41]:
```suggestion
      node_group:  # The subset of nodes for which to limit heat imports
```

File: doc/_static/custom_math/net_import_share.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        - expression: "[a, c]"
+      carrier:
```
Comment (sjpfenninger) [inline doc/_static/custom_math/net_import_share.yaml:43]:
```suggestion
      carrier:  # The carrier for which to limit imports
```

File: doc/_static/custom_math/net_import_share.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      - expression: net_import_share * sum(flow_out[techs=$transmission_techs, carriers=electricity], over=[techs, timesteps]) <= $total_energy_balance
+    sub_expressions:
+      total_energy_balance:
+        - expression: sum(flow_out[carriers=electricity], over=[techs, timesteps]) - sum(flow_in[carriers=electricity], over=[techs, timesteps])
+    slices:
+      transmission_techs:
+        - expression: get_transmission_techs(test_transmission_elec)
+
+  net_annual_import_share_max_node_group:
+    description: Limit upper bound on annual heat imports across two nodes in the model as a share of combined heat flows in both nodes.
```
Comment (sjpfenninger) [inline doc/_static/custom_math/net_import_share.yaml:32]:
```suggestion
    description: Limit upper bound on annual heat imports across a subset of nodes in the model as a share of combined heat flows in those nodes.
```

File: doc/_static/custom_math/net_import_share.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# Force upper limit on carrier imports within nodes or across all nodes as a share of all carrier flows at a node/across nodes.
+# Imports of transmission technologies are defined by `flow_out` (outflow from a transmission technology that originated in a remote node).
+
```
Comment (sjpfenninger) [inline doc/_static/custom_math/net_import_share.yaml:3]:
```suggestion
# We assume that transmission technologies `test_transmission_elec` and `test_transmission_heat` have been defined.

```

File: doc/_static/custom_math/net_import_share.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    foreach: [nodes, timesteps]
+    where: "defined(techs=test_transmission_elec, within=nodes, how=any)"
+    equations:
+      - expression: net_import_share * sum(flow_out[techs=$transmission_techs, carriers=electricity], over=techs) <= $total_energy_balance
+    sub_expressions:
+      total_energy_balance:
+        - expression: sum(flow_out[carriers=electricity], over=techs) - sum(flow_in[carriers=electricity], over=techs)
+    slices:
+      transmission_techs:
+        - expression: get_transmission_techs(test_transmission_elec)
```
Comment (sjpfenninger) [inline doc/_static/custom_math/net_import_share.yaml:16]: Once helper functions like `get_transmission_techs` are documented elsewhere, those docs could be linked to from here

**Review by sjpfenninger — CHANGES_REQUESTED** [review]

File: doc/_static/custom_math/net_import_share.yaml (hunk 5 above)
Comment (brynpickering) [reply doc/_static/custom_math/net_import_share.yaml:16]: They're already documented in the [API docs](https://calliope.readthedocs.io/en/latest/api/api.html#module-calliope.backend.helper_functions), although that is quite messy so I was planning to also give a run-down of them in the math section I'm writing in #423.

**Review by sjpfenninger — APPROVED** [review]

**Pull request merged**